### ISSUE TITLE
Overriding logging properties like FILE_LOG_PATTERN

### ISSUE THREAD

**Issue by marceloverdijk** [spring-projects/spring-boot#1788]
Feature request for adding support for overriding logging properties via application.properties/yaml.

E.g. I want to easily override the FILE_LOG_PATTERN property.
To define this in the application.properties/yaml also gives the benefit that I can change it per environment.

Boot already offers 

```
# LOGGING
logging.path=/var/logs
logging.file=myapp.log
logging.config= # location of config file (default classpath:logback.xml for logback)
logging.level.*= # levels for loggers, e.g. "logging.level.org.springframework=DEBUG" (TRACE, DEBUG, INFO, WARN, ERROR, FATAL, OFF)
```

Something like below maybe?

```
logging.property.FILE_LOG_PATTERN=
```

Reference: http://stackoverflow.com/questions/26629197/overriding-the-file-log-pattern-if-possible-per-env

**Comment by larry13767** [comment]
+1
agreed, this is a good feature.

**Comment by blackolo** [comment]
+1
nice to have.

**Comment by brenuart** [comment]
+1
nice and avoids having to duplicate all of the standard logback.xml configuration files provided with SpringBoot.

**Comment by cmoro-deusto** [comment]
+1 thanks very much.

**Comment by txiong1987** [comment]
+1 Thanks!

**Comment by edudar** [comment]
+1
Can you go beyond just `FILE_LOG_PATTERN` here? I suspect this duplicates what @marceloverdijk asked above to some extend. We have custom Logback appender that talks to `fluentd`. Nothing fancy, new xml that include base one and has its own appender. There are 3 properties though: `tag`, `host` and `port`. We can configure them now through `-D` option but it would be nice to have streamlined way to define logging properties that will be replaced in the same way `LOG_FILE` with `logging.file` work.

**Comment by igler** [comment]
+1

**Comment by cannibalcow** [comment]
+1

**Comment by nagase2** [comment]
I also struggle with same issue because the log message issued twice.
Eventually, I stop including base.xml at this
moment...
+1

**Comment by Chumper** [comment]
:+1: the same here, i want to use the `AmqpAppender` from spring-amqp: docs.spring.io/spring-amqp/docs/1.5.0.M1/api/org/springframework/amqp/rabbit/logback/AmqpAppender.html

The problem is, that i need to get the `host`, `username`, `password` and `port` from the application config.

**Comment by iamiddy** [comment]
+1 need this feature, 
opened this [SO](http://stackoverflow.com/questions/30580819/spring-environment-to-system-properties-for-logback-xml) before stumbled here

**Comment by hmeerlo** [comment]
+1

**Comment by ofbizbrazil** [comment]
Missing it a lot... When will it be done?

**Comment by xfcjscn** [comment]
+1

**Comment by spencergibb** [comment]
+1 for [spring-cloud-sleuth-correlation](https://github.com/spring-cloud-incubator/spring-cloud-sleuth/tree/master/spring-cloud-sleuth-correlation) including a correlation id in the logs.

**Comment by larry13767** [comment]
What is spring cloud sleuth correlation?  I am intrigued

**Comment by spencergibb** [comment]
@larry13767 see #2343 for the backstory.  Some of the code has been donated from https://github.com/4finance/micro-infra-spring  It's an attempt to provide tracing from the original request throughout the system regardless of what machine sub-requests are made on.

**Comment by rob-baily** [comment]
I was looking at ![a question](https://github.com/lukashinsch/spring-boot-extended-logging-properties) related to this request and it seems to me that the request to put them in the logging properties configuration may not be right.  Spring Boot is supposed to be logging implementation agnostic and it looks like we want to duplicate what would be Logback configuration in the logging proprieties space. @philwebb Have you seen ![this code](https://github.com/lukashinsch/spring-boot-extended-logging-properties)?  This seems like it might be a reasonable methodology to get some environment properties into the logging configuration and not have to replicate all logging configuration.  Otherwise if you think that adding some Logback specific properties makes sense I could maybe give it a shot unless someone else is already looking at it and you have some idea on what the properties should be named.

**Comment by ofbizbrazil** [comment]
any news about it?

**Comment by htynkn** [comment]
+1 any news?

**Comment by snicoll** [comment]
hey guys, this issue is scheduled for 1.3 (with a candidate on M2) so work on it should start soon. A PR would certainly speed things up!

**Comment by lukashinsch** [comment]
I gave it a shot last night (#3332), but I'm still struggling to get the CI build green (seems to be an environment dependent issue). In the meantime, I'd appreciate any feedback if the approach generally leads in the right direction.

**Comment by philwebb** [comment]
For anyone watching this issue there is now a new `<springProperty>` tag that you can use to pull in `Environment` properties. To use it you first need to rename your `logback.xml` file to `logback-spring.xml`. See the [updated documentation](http://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#boot-features-logback-extensions) for details. There's also a new `<springProfile>` tag that can be used to optionally include/exclude configuration.

I'd appreciate any feedback on this approach, does it work for people?

**Comment by ccit-spence** [comment]
How does this differ from using Spring Cloud Config?  A couple of months ago I needed the functionality for determining which profile was loaded and was able to accomplish with Spring Cloud Config.  Essentially loading profile specific logback configurations.  Is this the same concept?

**Comment by philwebb** [comment]
@ccit-spence can you point me at a sample to show me what you mean?

**Comment by ccit-spence** [comment]
@philwebb I didn't have to really add anything except for configuring the service to use Spring Cloud Config It was really the only thing I have to do. In each service I have the following logback profile specific files.  I do not have a logback.xml file.  Only with its respective profile as below.

``` 
logback-local.xml
logback-docker.xml
logback-production.xml
```
Your example seems to add more robust capabilities.  I am pretty sure the reason why Spring Cloud Config worked is that it refreshes the logback config after the initial load allowing for later processing during boot.

**Comment by philwebb** [comment]
@spencergibb @dsyer How does Spring Cloud Config support these profile specific logback configs? Is this something we should bring into Spring Boot? How does it deal with multiple active profiles?

**Comment by dsyer** [comment]
IIRC there is some hackery in Sprung Cloud Commons to explicitly give log configuration a chance to refresh before starting to main application context. Remember that Spring Cloud always has a parent "bootstrap" context and it is there that you can set up things that need extra logic before the main context has started. I'm not sure there is anything magic about it in general, but logging systems require special treatment to get them initialized early, so I think that led to a special case.

**Comment by rob-baily** [comment]
@philwebb I like the approach as it seems very flexible.  Based on this is it possible to update properties that are defined in defaults.xml like the CONSOLE_LOG_PATTERN and FILE_LOG_PATTERN from another property source?  It seems like it would be useful to override just these and not completely define a new configuration just to override this.

**Comment by philwebb** [comment]
@rob-baily I've raised #3367 to consider this. We can't use the same `<springProperty/>` approach since `base.xml` might be imported from from a standard logback config where we don't have the extensions.

**Comment by rakesh9** [comment]
I am trying to use  springProfile tag in logback.xml to have profile specific logging but spring boot is ignoring this and all tags after this.(I am using spring boot version 1.2.4 Release)
-ERROR in ch.qos.logback.core.joran.spi.Interpreter@5:47 - no applicable action for [springProfile], current ElementPath  is [[configuration][springProfile]]
-ERROR in ch.qos.logback.core.joran.spi.Interpreter@6:88 - no applicable action for [appender], current ElementPath  is [[configuration][springProfile][appender]]
--------------
I tried putting the same configuration in logback-spring.xml but spring boot not recognizing and loading this file itself.
Please guide me how to use  springProfile tag to get profile specific logging.

**Comment by rob-baily** [comment]
@rakesh9 The new feature is coming is version 1.3.0.  You can't use it until this unless you want to go against a SNAPSHOT build.

**Comment by philwebb** [comment]
@rakesh9 as @rob-baily said this is a new feature not and won't be available in 1.2.x. You'll also need to use `logback-spring.xml` as your filename when you do migrate (see the [updated docs](http://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#boot-features-logback-extensions)).

**Comment by rakesh9** [comment]
@rob-baily @ philwebb  Thanks for the update..I missed observing the version...I am new to spring-boot..
can you suggest me any best alternative to have conditional logging for diff environments like dev,test using logback.groovy in 1.2.4 version

**Comment by edward-smith** [comment]
@philwebb Love the the new `<springProfile>` element in 1.3.0.M2, though I had a question:

The `name` attribute of `<springProfile>` doesn't seem to accept a list of profiles (I've tried with spaces, commas, semi-colons), is this working as intended?

**Comment by philwebb** [comment]
@edward-smith It only accepts a single name at the moment, but it should be easy to fix that. I'll open a new issue.

**Comment by thefalconfeat** [comment]
+1

**Comment by snicoll** [comment]
@thefalconfeat it's already implemented (and this issue is closed FWIW).

**Comment by lhanson** [comment]
Is there any support for this functionality in `logback-spring.groovy`?

**Comment by wilkinsona** [comment]
@lhanson It's XML only. We're not sure that it's possible with Logback's Groovy support. See #5051 for details.

**Comment by lhanson** [comment]
@wilkinsona Great, thanks. I think I'll be able to express my setup in XML, so this feature is still a big help.

**Comment by Tetradeus** [comment]
Any chance to see this feature (springproperty) extended for log4j2 ?

**Comment by wilkinsona** [comment]
@Tetradeus We're not sure that it's possible with Log4j2. If it is, the change would be made as part of #4032.

**Comment by mageddo** [comment]
I want to know where Spring specify that spring boot 1.2 not support log pattern set by application.properties, man, I take a while to get it.

**Comment by wilkinsona** [comment]
The string `FILE_LOG_PATTERN` doesn't appear anywhere in the documentation for 1.2, whereas it appears in the documentation for 1.3 and onwards. I'm not sure that there's much more that we can do.
